test/docker/presync-webhook-server/oci-signature-verification-server.go

```diff
@@ -0,0 +1,154 @@
+	if image == "" {
+		return nil
+	}
+
+	pubKey, err := signature.LoadPublicKey(ctx, publicKeyPath)
+	if err != nil {
+		return fmt.Errorf("error loading public key: %v", err)
+	}
+
+	googleAuth, err := google.NewEnvAuthenticator(ctx)
+	if err != nil {
+		return err
+	}
+
+	opts := &cosign.CheckOpts{
+		RegistryClientOpts: []ociremote.Option{ociremote.WithRemoteOptions(remote.WithAuth(googleAuth))},
+		SigVerifier:        pubKey,
+		IgnoreTlog:         true,
+	}
+
+	ref, err := name.ParseReference(image)
+	if err != nil {
+		return fmt.Errorf("failed to parse image reference: %v", err)
+	}
+	_, _, err = cosign.VerifyImageSignatures(ctx, ref, opts)
+	if err != nil {
+		return fmt.Errorf("image verification failed for %s: %v", image, err)
+	}
+
+	klog.Infof("Image %s verified successfully", image)
+	return nil
+}
+
+func handleWebhook(w http.ResponseWriter, r *http.Request) {
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		klog.Errorf("Failed to read request body: %v", err)
+		http.Error(w, "Failed to read request body", http.StatusBadRequest)
+		return
+	}
+
+	var admissionReview admissionv1.AdmissionReview
+	if err := json.Unmarshal(body, &admissionReview); err != nil {
+		klog.Errorf("Failed to unmarshal admission review: %v", err)
+		http.Error(w, "Failed to unmarshal admission review", http.StatusBadRequest)
+		return
+	}
+
+	response := &admissionv1.AdmissionResponse{
+		UID: admissionReview.Request.UID,
+	}
+
+	oldAnnotations, err := getAnnotations(admissionReview.Request.OldObject.Raw)
+	if err != nil {
+		klog.Errorf("Failed to extract old annotations: %v", err)
+		response.Result = &metav1.Status{
+			Message: fmt.Sprintf("Failed to extract old annotations: %v", err),
+		}
+		response.Allowed = false
+	}
+
+	newAnnotations, err := getAnnotations(admissionReview.Request.Object.Raw)
+	if err != nil {
+		klog.Errorf("Failed to extract new annotations: %v", err)
+		response.Result = &metav1.Status{
+			Message: fmt.Sprintf("Failed to extract new annotations: %v", err),
+		}
+		response.Allowed = false
+	}
+
+	if newAnnotations[imageToSync] != oldAnnotations[imageToSync] {
+		klog.Infof("Annotation image-to-sync changed from %s to %s", oldAnnotations[imageToSync], newAnnotations[imageToSync])
+		if err := verifyImageSignature(r.Context(), newAnnotations[imageToSync]); err != nil {
+			klog.Errorf("Image verification failed: %v", err)
+			response.Allowed = false
+			response.Result = &metav1.Status{
+				Message: fmt.Sprintf("Image verification failed: %v", err),
+			}
+		} else {
+			klog.Infof("Image verification successful for %s", newAnnotations[imageToSync])
+			response.Allowed = true
+		}
+	} else {
+		response.Allowed = true
+	}
+
+	admissionReview.Response = response
+	if err := json.NewEncoder(w).Encode(admissionReview); err != nil {
+		klog.Errorf("Failed to encode admission response: %v", err)
+	}
+}
+
+func main() {
+	http.HandleFunc("/validate", handleWebhook)
+	klog.Info("Starting webhook server on port 8443...")
```

Comment:
I explained why we need to use `8443` instead of `443` in a commit message. It's better to add an inline comment here.

// Use 8443 as a non-privileged port to avoid root privilege to bind to the port

Reply:
Done